File: Java/Tree/Tree_07.java
/* 
=============================================================================== 
Program Name    : Binary Search Tree - Count Total Nodes 
Program Number  : 07 
File Name       : Tree_07.java 
 
Description     : This program demonstrates how to count the total number 
                  of nodes in a Binary Search Tree using Java. It uses 
                  recursion to visit every node in the tree and calculates 
                  the total number of nodes. 
 
Date            : August 2026 
Language        : Java 
Author          : Ishwari Vijaykumar Surve 
=============================================================================== 
 
Concepts Used: 
1. Class 
2. Object-Oriented Programming (OOP) 
3. Self-Referential Class 
4. Constructor 
5. Binary Search Tree (BST) 
6. Dynamic Memory Allocation 
7. Recursion 
8. Insertion Operation 
9. Node Counting 
10. Function Overloading 
 
Key Learning Points: 
- Understand how to count total nodes in a Binary Search Tree. 
- Learn how recursion is used to count nodes. 
- Return 0 when the current node is NULL. 
- Count nodes in the left subtree. 
- Count nodes in the right subtree. 
- Add 1 for the current node. 
- Understand function overloading using two Count functions. 
 
=============================================================================== 
*/ 
 
class node 
{ 
    public int data; 
    public node lchild; 
    public node rchild; 
} 
 
class Tree 
{ 
    private node first; 
 
    public Tree() 
    { 
        first = null; 
    } 
 
    /////////////////////////////////////////////////////////////////////////// 
    // 
    // Function Name : Insert 
    // Description   : Insert element into Binary Search Tree 
    // 
    /////////////////////////////////////////////////////////////////////////// 
 
    public void Insert(int iNo) 
    { 
        node newn = null; 
        node temp = null; 
 
        newn = new node(); 
 
        newn.data = iNo; 
        newn.lchild = null; 
        newn.rchild = null; 
 
        if(first == null) 
        { 
            first = newn; 
        } 
        else 
        { 
            temp = first; 
 
            while(true) 
            { 
                if(iNo > temp.data) 
                { 
                    if(temp.rchild == null) 
                    { 
                        temp.rchild = newn; 
                        break; 
                    } 
 
                    temp = temp.rchild; 
                } 
                else if(iNo < temp.data) 
                { 
                    if(temp.lchild == null) 
                    { 
                        temp.lchild = newn; 
                        break; 
                    } 
 
                    temp = temp.lchild; 
                } 
                else 
                { 
                    System.out.println("Duplicate element not allowed"); 
                    break; 
                } 
            } 
        } 
    } 
 
    /////////////////////////////////////////////////////////////////////////// 
    // 
    // Function Name : Count 
    // Description   : Count total number of nodes 
    // 
    /////////////////////////////////////////////////////////////////////////// 
 
    public int Count(node temp) 
    { 
        if(temp == null) 
        { 
            return 0; 
        } 
 
        return Count(temp.lchild) + Count(temp.rchild) + 1; 
    } 
 
    /////////////////////////////////////////////////////////////////////////// 
    // 
    // Function Name : Count 
    // Description   : Start counting from first node 
    // 
    /////////////////////////////////////////////////////////////////////////// 
 
    public int Count() 
    { 
        return Count(first); 
    } 
} 
 
class Tree_07 
{ 
    public static void main(String A[]) 
    { 
        Tree tobj = new Tree(); 
 
        tobj.Insert(11); 
        tobj.Insert(5); 
        tobj.Insert(17); 
        tobj.Insert(4); 
        tobj.Insert(7); 
        tobj.Insert(15); 
        tobj.Insert(21); 
 
        int iRet = tobj.Count(); 
 
        System.out.println("Total Number of Nodes : " + iRet); 
    } 
} 
 
/* 
=============================================================================== 
Expected Output: 
=============================================================================== 
 
Total Number of Nodes : 7 
 
=============================================================================== 
Memory Representation: 
=============================================================================== 
 
                         11 
                       /    \ 
                      5      17 
                     / \    /  \ 
                    4   7  15  21 
 
Total Nodes: 
 
             11 
           /    \ 
          5      17 
         / \    /  \ 
        4   7  15  21 
 
        1 + 1 + 1 + 1 + 1 + 1 + 1 = 7 
 
Total Number of Nodes = 7 
 
=============================================================================== 
Counting Process: 
=============================================================================== 
 
Step 1 : Start from the root node 11. 
Step 2 : Count all nodes in the left subtree of 11. 
Step 3 : Count node 5 and its child nodes 4 and 7. 
Step 4 : Left subtree contains 3 nodes. 
Step 5 : Count all nodes in the right subtree of 11. 
Step 6 : Count node 17 and its child nodes 15 and 21. 
Step 7 : Right subtree contains 3 nodes. 
Step 8 : Add the root node 11. 
 
Total Nodes = Left Subtree + Right Subtree + Root 
            = 3 + 3 + 1 
            = 7 
 
=============================================================================== 
Node Structure: 
=============================================================================== 
 
                    +-----------+-----------+-----------+ 
                    |   data    |  lchild   |  rchild   | 
                    +-----------+-----------+-----------+ 
                          |           |           | 
                        Value       Address     Address 
 
=============================================================================== 
Algorithm: 
=============================================================================== 
 
1. Start the program. 
2. Define the node class with data, lchild and rchild members. 
3. Define the Tree class. 
4. Initialize first to NULL using the constructor. 
5. Define the Insert function. 
6. Insert the given elements into the Binary Search Tree. 
7. Define the recursive Count function. 
8. Check whether the current node is NULL. 
9. If the current node is NULL, return 0. 
10. Recursively count nodes in the left subtree. 
11. Recursively count nodes in the right subtree. 
12. Add 1 for the current node. 
13. Return the total count. 
14. Call Count starting from the first node. 
15. Display the total number of nodes. 
16. Stop the program. 
 
=============================================================================== 
Program Flow: 
=============================================================================== 
 
        Start 
          | 
          v 
     Create Tree 
          | 
          v 
   Insert Tree Elements 
          | 
          v 
      Call Count() 
          | 
          v 
      Start at Root 
          | 
          v 
      Is Node NULL? 
        /       \ 
      Yes        No 
       |          | 
       v          v 
   Return 0   Count Left 
                  | 
                  v 
             Count Right 
                  | 
                  v 
              Add 1 
                  | 
                  v 
           Return Count 
                  | 
                  v 
           Display Result 
                  | 
                  v 
                 Stop 
 
=============================================================================== 
Time Complexity: 
=============================================================================== 
 
Insertion   : O(log n) Average 
Insertion   : O(n) Worst Case 
Count       : O(n) 
 
=============================================================================== 
Space Complexity: 
=============================================================================== 
 
Tree Storage    : O(n) 
Recursion Stack : O(h) 
 
Where h is the height of the Binary Search Tree. 
 
=============================================================================== 
Conclusion: 
=============================================================================== 
 
Program 07 demonstrates how to count the total number of nodes in a Binary 
Search Tree using Java. The recursive Count function counts the nodes in 
the left subtree, counts the nodes in the right subtree and adds one for 
the current node. For the given Tree, the total number of nodes is 7. 
 
=============================================================================== 
*/

File: C++/Tree/Tree_07.cpp
/*
===============================================================================
Program Name    : Binary Search Tree - Count Total Nodes
Program Number  : 07
File Name       : Tree_07.cpp 
Description     : This program counts the total number of nodes present in
                  a Binary Search Tree using recursive traversal.

Date            : August 2026
Language        : C++
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Binary Search Tree (BST)
2. Recursion
3. Tree Traversal
4. Class and Object
5. Self-Referential Structure
6. Dynamic Memory Allocation
7. Pointers

Key Learning Points:
- Understand how to count all nodes in a Binary Search Tree.
- Learn recursive traversal of a Tree.
- Understand the base condition of recursion.
- Learn how results from left and right subtrees are combined.
- Understand the use of a wrapper function.

===============================================================================
*/

#include<iostream>
using namespace std;

#pragma pack(1)

struct node
{
    int data;
    struct node *lchild;
    struct node *rchild;
};

class BST
{
private:
    struct node *first;

public:
    BST();
    void Insert(int iNo);
    int Count(struct node *first);
    int CountNode();
};

// Constructor
BST::BST()
{
    first = NULL;
}

// Insert Node
void BST::Insert(int iNo)
{
    struct node *newn = NULL;
    struct node *temp = NULL;

    newn = new node;

    newn->data = iNo;
    newn->lchild = NULL;
    newn->rchild = NULL;

    if(first == NULL)
    {
        first = newn;
    }
    else
    {
        temp = first;

        while(1)
        {
            if(iNo > temp->data)
            {
                if(temp->rchild == NULL)
                {
                    temp->rchild = newn;
                    break;
                }

                temp = temp->rchild;
            }
            else if(iNo < temp->data)
            {
                if(temp->lchild == NULL)
                {
                    temp->lchild = newn;
                    break;
                }

                temp = temp->lchild;
            }
            else
            {
                cout<<"Duplicate element not allowed\n";
                delete newn;
                break;
            }
        }
    }
}

// Count Total Nodes
int BST::Count(struct node *first)
{
    if(first == NULL)
    {
        return 0;
    }

    return Count(first->lchild) +
           Count(first->rchild) + 1;
}

// Wrapper Function
int BST::CountNode()
{
    return Count(first);
}

int main()
{
    BST bobj;
    int iRet = 0;

    bobj.Insert(11);
    bobj.Insert(5);
    bobj.Insert(17);
    bobj.Insert(4);
    bobj.Insert(7);
    bobj.Insert(15);
    bobj.Insert(21);

    iRet = bobj.CountNode();

    cout<<"Total Nodes : "<<iRet<<endl;

    return 0;
}

/*
===============================================================================
Expected Output:
===============================================================================

Total Nodes : 7

===============================================================================
Memory Representation:
===============================================================================

                         first
                           |
                           ▼
                       +-------+
                       |  11   |
                       +---+---+
                          / \
                         /   \
                        ▼     ▼
                    +-----+ +-----+
                    |  5  | | 17  |
                    +--+--+ +--+--+
                      / \     / \
                     /   \   /   \
                    ▼     ▼ ▼     ▼
                  +---+ +---+ +---+ +---+
                  | 4 | | 7 | |15 | |21 |
                  +---+ +---+ +---+ +---+

Total Nodes = 7

===============================================================================
Algorithm:
===============================================================================

1. Start the program.
2. Create a Binary Search Tree object.
3. Insert the required elements into the BST.
4. Call CountNode().
5. CountNode() calls Count() with the root node.
6. Check whether the current node is NULL.
7. If the node is NULL, return 0.
8. Recursively count the nodes in the left subtree.
9. Recursively count the nodes in the right subtree.
10. Add 1 for the current node.
11. Return the total count.
12. Display the total number of nodes.
13. Stop the program.

===============================================================================
Program Flow:
===============================================================================

                         Start
                           │
                           ▼
                    Create BST Object
                           │
                           ▼
                    Insert BST Nodes
                           │
                           ▼
                   Call CountNode()
                           │
                           ▼
                     Call Count()
                           │
                           ▼
                     Is Node NULL?
                      /          \
                    Yes           No
                    │              │
                    ▼              ▼
                  Return 0    Count Left Subtree
                                   │
                                   ▼
                             Count Right Subtree
                                   │
                                   ▼
                          Add 1 for Current Node
                                   │
                                   ▼
                            Return Total Count
                                   │
                                   ▼
                           Display Total Nodes
                                   │
                                   ▼
                                  Stop

===============================================================================
Time Complexity:
===============================================================================

Insert() - Best Case    : O(1)
Insert() - Average Case : O(log n)
Insert() - Worst Case   : O(n)

Count() - Best Case     : O(n)
Count() - Average Case  : O(n)
Count() - Worst Case    : O(n)

Where n is the number of nodes in the BST.

===============================================================================
Space Complexity:
===============================================================================

Count() Auxiliary Space : O(h)

Where h is the height of the Tree.

For a balanced Tree : O(log n)
For a skewed Tree   : O(n)

Total Space : O(n)

===============================================================================
Conclusion:
===============================================================================

Program 007 counts the total number of nodes present in a Binary Search
Tree using recursion. The Count() function recursively visits the left
and right subtrees and adds one for the current node. This demonstrates
how recursive Tree traversal can be used to calculate Tree properties.

===============================================================================
*/
